.azure-pipelines/impacted_area_testing/calculate_instance_number.py

```diff
@@ -0,0 +1,133 @@
+    az_login_cmd = f"az login --identity --username {managed_identity_id}"
+    az_login_attempts = 0
+    while az_login_attempts < MAX_GET_TOKEN_RETRY_TIMES:
+        try:
+            result = os.popen(az_login_cmd)
+            result.read()
+            break
+        except Exception as exception:
+            az_login_attempts += 1
+            raise Exception(
+                f"Failed to az login with exception: {repr(exception)}. "
+                f"Retry {MAX_GET_TOKEN_RETRY_TIMES - az_login_attempts} times to login."
+            )
+
+    # If az login failed, return with exception
+    if az_login_attempts >= MAX_GET_TOKEN_RETRY_TIMES:
+        raise Exception(f"Failed to az login after {MAX_GET_TOKEN_RETRY_TIMES} attempts.")
+
+    # 2. Get access token with re-try
+    get_token_cmd = "az account get-access-token --resource https://api.kusto.windows.net --query accessToken -o tsv"
+    get_token_attempts = 0
+    while get_token_attempts < MAX_GET_TOKEN_RETRY_TIMES:
+        try:
+            result = os.popen(get_token_cmd)
+            access_token = result.read()
+            if not access_token:
+                raise Exception("Parse token from stdout failed, accessToken is None.")
+
+            return access_token
+
+        except Exception as exception:
+            get_token_attempts += 1
+            raise Exception(f"Failed to get token with exception: {repr(exception)}.")
+
+    # If az get token failed, return with exception
+    if get_token_attempts >= MAX_GET_TOKEN_RETRY_TIMES:
+        raise Exception(f"Failed to get token after {MAX_GET_TOKEN_RETRY_TIMES} attempts")
+
+
+def main(scripts, topology, branch):
+    ingest_cluster = os.getenv("TEST_REPORT_QUERY_KUSTO_CLUSTER_BACKUP")
+    access_token = get_access_token()
+
+    if not ingest_cluster or not access_token:
+        raise RuntimeError(
+            "Could not load Kusto Credentials from environment")
+
+    try:
+        kcsb = KustoConnectionStringBuilder.with_aad_application_token_authentication(ingest_cluster,
+                                                                                      access_token)  # noqa F841
+        client = KustoClient(kcsb)
+    except Exception as e:
+        raise Exception("Connect to kusto fails, error {}".format(e))
+
+    scripts = parse_list_from_str(scripts)
+
+    scripts_running_time = {}
+    total_running_time = 0
+
+    for script in scripts:
+        # As baseline test is the universal set of PR test
+        # we get the historical running time of one script here
+        # We get recent 5 test plans and calculate the average running time
+        query = "V2TestCases " \
+                "| join kind=inner" \
+                "(TestPlans " \
+                "| where TestPlanType == 'PR' and Result == 'FINISHED' and Topology == '{}' " \
+                "and TestBranch == '{}' and TestPlanName contains '{}' " \
+                "and TestPlanName contains '_BaselineTest_'" \
```

Comment:
Baseline only applied on master branch, it would cause empty testplans for other branches, so I suggest to filter recent 5 success testcases, not success testplans

Reply:
In table `V2TestCases`, there is too little information. Here, we must filiter out the test case which has the same branch and topology. To get these information, we must join the table `TestPlans`.

Reply:
I have 2 points here:
1. baseline only applied for master branch, so your query would not get any result for other branches like 202405, so I suggest to remove condition TestPlanName contains '_BaselineTest_'
2. In your query, if recent 5 test results are "failure", it's not accurate, so I suggest to add another filter "where result !in ("failure", "error") in V2TestCases table

Reply:
For the first question, 
We have many other testplans except for baseline test. Here, we need to use the result of baseline test, so the condition is necessary.

Reply:
For the second one, the `Result == 'FINISHED'` for testplan can't represent the success of all test scipts in this testplan?

Reply:
1. Dose TestPlanType == 'PR' and CreatedByType == 'PR' not enough to filter them? I suppose there won't be much difference, or you can limit results with a time range. Impact area feature would extend to other branches even internal, baseline may not cover such a range of branches, it's only a matter of time to make this change.
2. testplan result == 'FINISHED' doesn't mean testcase result not in ("failure", "error"), PR test would retry, baseline would not stop in failure

Reply:
> 1. Impact area feature would extend to other branches even internal, baseline may not cover such a range of branches, it's only a matter of time to m

1. No, there are other kinds of PR testing, like created by pull request in sonic-mgmt/sonic-buildimage, and some manually triggered PRs. 
2. I will add this condition.